plutus-core/plutus-core/src/Crypto/BLS12_381/Pairing.hs

```diff
@@ -0,0 +1,64 @@
+{-# LANGUAGE DeriveAnyClass        #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE TypeApplications      #-}
+module Crypto.BLS12_381.Pairing
+    (
+     MlResult (..),
+     millerLoop,
+     mulMlResult,
+     finalVerify,
+     mlResultMemSizeBytes
+    ) where
+
+import Crypto.External.EllipticCurve.BLS12_381 qualified as BlstBindings
+import Crypto.External.EllipticCurve.BLS12_381.Internal qualified as BlstBindings.Internal
+
+import Crypto.BLS12_381.G1 qualified as G1
+import Crypto.BLS12_381.G2 qualified as G2
+import PlutusCore.Pretty.PrettyConst (ConstConfig)
+import Text.PrettyBy (PrettyBy, prettyBy)
+
+import Control.DeepSeq (NFData, rnf)
+import Data.Bifunctor (second)
+import Flat
+import Prettyprinter
+
+{- | This type represents the result of computing a pairing using the Miller
+   loop.  Values of this type are ephemeral, only created during script
+   execution.  We do not provide any means of serialising, deserialising,
+   printing, or parsing MlResult values. -}
+newtype MlResult = MlResult { unMlResult :: BlstBindings.PT }
+    deriving newtype (Eq)
+instance Show MlResult where
+    show _ = "<opaque>"
+instance Pretty MlResult where
+    pretty = pretty . show
+instance PrettyBy ConstConfig MlResult where
+    prettyBy _ = pretty
+-- We need a Flat instance to get everything to build properly; however we'll
+-- never want MlResult values in serialised scripts, so the decoding and
+-- encoding functions just raise errors.
+instance Flat MlResult where
+    decode = fail "BLS12_381.Pairing.MlResult: flat decoding disallowed"
+    encode = error "BLS12_381.Pairing.MlResult: flat encoding disallowed"
```

Comment:
I'm not sure what to do there.  I think that the only way this could arise is if someone creates an `MlResult` value in Haskell and then tries to serialise it (we don't have any concrete syntax for constants of this type, and we don't provide any serialisation method in PLC). Maybe it's better to issue an error as early as possible to tell the user that they're doing something silly rather than silently serialising a fake value and then exploding at deserialisation time.